libraries/AP_NMEA_Output/AP_NMEA_Output.cpp

```diff
@@ -0,0 +1,178 @@
+    struct tm* tm = gmtime(&time_sec);
+
+    // format time string
+    char tstring[11];
+    snprintf(tstring, sizeof(tstring), "%02u%02u%06.3f", tm->tm_hour, tm->tm_min, tm->tm_sec + (time_usec % 1000000) * 1.0e-6);
+
+    // format date string
+    char dstring[7];
+    snprintf(dstring, sizeof(dstring), "%02u%02u%02u", tm->tm_mday, tm->tm_mon+1, tm->tm_year % 100);
+
+    AP_AHRS_NavEKF& ahrs = AP::ahrs_navekf();
+
+    // get location (note: get_position from AHRS always returns true after having GPS position once)
+    Location loc;
+    bool pos_valid = ahrs.get_location(loc);
+
+    // format latitude
+    char lat_string[13];
+    float deg = fabsf(loc.lat * 1.0e-7f);
+    snprintf(lat_string,
+             sizeof(lat_string),
+             "%02u%08.5f,%c",
+             (unsigned) deg,
+             double((deg - int(deg)) * 60),
+             loc.lat < 0 ? 'S' : 'N');
+
+    // format longitude
+    char lng_string[14];
+    deg = fabsf(loc.lng * 1.0e-7f);
+    snprintf(lng_string,
+             sizeof(lng_string),
+             "%03u%08.5f,%c",
+             (unsigned) deg,
+             double((deg - int(deg)) * 60),
+             loc.lng < 0 ? 'W' : 'E');
+
+    // format GGA message
+    char* gga = nullptr;
+    int16_t gga_res = asprintf(&gga,
+                               "$GPGGA,%s,%s,%s,%01d,%02d,%04.1f,%07.2f,M,0.0,M,,",
+                               tstring,
+                               lat_string,
+                               lng_string,
+                               pos_valid ? 1 : 0,
+                               pos_valid ? 6 : 3,
+                               2.0,
+                               loc.alt * 0.01f);
+    char gga_end[6];
+    snprintf(gga_end, sizeof(gga_end), "*%02X\r\n", (unsigned) _nmea_checksum(gga));
+
+    // get speed
+    Vector2f speed = ahrs.groundspeed_vector();
+    float speed_knots = norm(speed.x, speed.y) * M_PER_SEC_TO_KNOTS;
+    float heading = wrap_360(degrees(atan2f(speed.x, speed.y)));
+
+    // format RMC message
+    char* rmc = nullptr;
+    int16_t rmc_res = asprintf(&rmc,
+                               "$GPRMC,%s,%c,%s,%s,%.2f,%.2f,%s,,",
+                               tstring,
+                               pos_valid ? 'A' : 'V',
+                               lat_string,
+                               lng_string,
+                               speed_knots,
+                               heading,
+                               dstring);
+    char rmc_end[6];
+    snprintf(rmc_end, sizeof(rmc_end), "*%02X\r\n", (unsigned) _nmea_checksum(rmc));
+
+    // send to all NMEA output ports
+    for (uint8_t i = 0; i < _num_outputs; i++) {
+        if (gga_res != -1) {
+            _uart[i]->write(gga);
+            _uart[i]->write(gga_end);
+        }
+
+        if (rmc_res != -1) {
+            _uart[i]->write(rmc);
+            _uart[i]->write(rmc_end);
+        }
+    }
+
+    if (gga_res != -1) {
+        free(gga);
```

Comment:
Could just free the pointers here.  If they're null then it's a noop.

Reply:
I ended up not doing this because `asnprint` documentation says that pointer value would be undefined if there is an error - I had probably seen that, I was surprised why I had done this, so I think that's the reason.